Add first_over_time and ts_of_first_over_time as experimental functions

ringerc opened the pull request:
promql: add `first_over_time(...)` and `ts_of_first_over_time(...)` functions per https://github.com/prometheus/prometheus/issues/16963

Test cover is added for the functions and documentation is updated. `make test-short` passes; I'll run the longer suite too.

The `first_over_time()` function is particularly useful when used in a range query as `first_over_time(m[step()])` with `--enable-feature=promql-duration-expr` to find the first sample within the step range. Using `offset step()` will instead select the newest sample _outside_ the range, if such a sample exists.


#### Which issue(s) does the PR fix:

https://github.com/prometheus/prometheus/issues/16963

#### Does this PR introduce a user-facing change?

```release-notes
[FEATURE] PromQL: Add `first_over_time(...)` and `ts_of_first_over_time(...)` behind feature flag `experimental-promql-functions #17021 #16963
```

ringerc commented:
I have a second patch here that adds another test case, but this case does not pass and I can't figure out why yet.
[0001-test-add-test-cover-for-series-with-histograms-follo.patch.txt](https://github.com/user-attachments/files/21648662/0001-test-add-test-cover-for-series-with-histograms-follo.patch.txt)


As far as I can tell, this data-load directive

```
load 10s
  # [...]
  data_histogram{type="mix_samples_hfirst"} {{schema:0 sum:1 count:2}} {{schema:0 sum:9 count:3}} 4 1
```

is behaving as if the histogram samples are entirely absent.

Notably, `count_over_time(data_histogram{type="mix_samples_hfirst"})` returns 2, not the expected 4. And the new `first_over_time(...)` returns the first float sample 4, not the first sample overall (a histogram sample `{{schema:0 sum:1 count:2}}`).

The fact that this affects `count_over_time`, which I have not changed, suggests either I've somehow mis-defined the sample vector, the test engine is somehow omitting them, or there's some problematic optimisation going on.

Both `count_over_time(...)` and the new `first_over_time(...)` correctly handle a mixed vector where the _last_ sample is a histogram. But they do not handle mixed float/histogram samples where the last sample is a float.

ringerc commented:
[make-test-out.txt](https://github.com/user-attachments/files/21674744/make-test-out.txt)

`make test` output

beorn7 commented:
@roidelapluie could you have a look? (I have it on my list, too, but it is a long list, and I would feel better if you checked this anyway.)

ringerc commented on promql/functions.go:
```diff
@@ -767,6 +767,34 @@ func funcCountOverTime(_ []Vector, matrixVals Matrix, _ parser.Expressions, enh
+
+	// If a histogram data point exists and is older than any float
+	// data points, return it.
+	if len(el.Histograms) > 0 && (len(el.Floats) == 0 || (h.T < f.T)) {
```
Updated to be the same as `last_over_time` (I rebased the change into the series), and rebased onto `main`

```
➜  prometheus git:(add_first_over_time) ✗ go test -v -run 'TestEvaluations/testdata/functions.test' ./promql |grep -i first_over
=== RUN   TestEvaluations/testdata/functions.test/line_1330/ts_of_first_over_time(metric[90s])
=== RUN   TestEvaluations/testdata/functions.test/line_1333/ts_of_first_over_time(metric[90s])
=== RUN   TestEvaluations/testdata/functions.test/line_1336/ts_of_first_over_time(metric[90s])
=== RUN   TestEvaluations/testdata/functions.test/line_1339/ts_of_first_over_time(metric_histogram{type="only_histogram"}[90s])
=== RUN   TestEvaluations/testdata/functions.test/line_1342/ts_of_first_over_time(metric_histogram{type="mix"}[90s])
=== RUN   TestEvaluations/testdata/functions.test/line_1611/first_over_time({__name__=~"data(_histogram|_sparse|_empty)?"}[2m])
        --- PASS: TestEvaluations/testdata/functions.test/line_1330/ts_of_first_over_time(metric[90s]) (0.00s)
        --- PASS: TestEvaluations/testdata/functions.test/line_1333/ts_of_first_over_time(metric[90s]) (0.00s)
        --- PASS: TestEvaluations/testdata/functions.test/line_1336/ts_of_first_over_time(metric[90s]) (0.00s)
        --- PASS: TestEvaluations/testdata/functions.test/line_1339/ts_of_first_over_time(metric_histogram{type="only_histogram"}[90s]) (0.00s)
        --- PASS: TestEvaluations/testdata/functions.test/line_1342/ts_of_first_over_time(metric_histogram{type="mix"}[90s]) (0.00s)
        --- PASS: TestEvaluations/testdata/functions.test/line_1611/first_over_time({__name__=~"data(_histogram|_sparse|_empty)?"}[2m]) (0.00s)
```

ringerc replied:
Also updated the PR template for the new (?) relnotes support

ringerc commented:
The CI test failed with

> --- FAIL: TestShutdown (14.28s)
>    queue_manager_test.go:496: Took too long to shutdown: 1.296696568s > 1s

This strikes me as likely worker load related or unrelated to the PR; certainly I've changed nothing relevant there. 

Would anybody mind re-running the test and having a look at the PR now I've addressed the comments in it?

ringerc commented:
@roidelapluie  Would you be willing to re-check this PR?

ringerc commented on docs/querying/functions.md:
```diff
@@ -849,7 +849,7 @@ over time and return an instant vector with per-series aggregation results:
 * `stddev_over_time(range-vector)`: the population standard deviation of all float samples in the specified interval.
 * `stdvar_over_time(range-vector)`: the population standard variance of all float samples in the specified interval.
-* `last_over_time(range-vector)`: the most recent sample in the specified interval.
+* `last_over_time(range-vector)`: the most recent sample in the specified interval, including NaN samples.
```
Because the docs didn't explain that NaN samples were counted, and to me it's unintuitive. I can revert that change and split it out if desired.

ringerc replied:
```
commit 64060430557ad95194d9393fd82b2107b810b162 (HEAD -> add_first_over_time)
Author: Craig Ringer <craig.ringer@enterprisedb.com>
Date:   Tue Sep 2 10:13:01 2025 +1200

    fixup! feat(promql): add first_over_time and ts_of_first_over_time

diff --git a/docs/querying/functions.md b/docs/querying/functions.md
index 1e380108b..4776490e5 100644
--- a/docs/querying/functions.md
+++ b/docs/querying/functions.md
@@ -849,7 +849,7 @@ over time and return an instant vector with per-series aggregation results:
 * `quantile_over_time(scalar, range-vector)`: the φ-quantile (0 ≤ φ ≤ 1) of all float samples in the specified interval.
 * `stddev_over_time(range-vector)`: the population standard deviation of all float samples in the specified interval.
 * `stdvar_over_time(range-vector)`: the population standard variance of all float samples in the specified interval.
-* `last_over_time(range-vector)`: the most recent sample in the specified interval, including NaN samples.
+* `last_over_time(range-vector)`: the most recent sample in the specified interval.
 * `present_over_time(range-vector)`: the value 1 for any series in the specified interval.
 
 If the [feature flag](../feature_flags.md#experimental-promql-functions)
@@ -863,11 +863,10 @@ additional functions are available:
 * `ts_of_max_over_time(range-vector)`: the timestamp of the last float sample
   that has the maximum value of all float samples in the specified interval.
 * `ts_of_last_over_time(range-vector)`: the timestamp of last sample in the
-  specified interval, including NaN samples
-* `first_over_time(range-vector)`: the oldest sample in the specified interval,
-   including NaN samples.
+  specified interval.
+* `first_over_time(range-vector)`: the oldest sample in the specified interval.
 * `ts_of_first_over_time(range-vector)`: the timestamp of earliest sample in the
-  specified interval, including NaN samples.
+  specified interval.
 
 Note that all values in the specified interval have the same weight in the
 aggregation even if the values are not equally spaced throughout the interval.
```
